juju/juju
# [JUJU-804] Remove facade reflection during creation.

Using reflection for every facade construction was a flex, that was dumb. 
Instead, we should be boring and create facades using functions that are 
typed by the compiler, so if we make any mistakes we don’t have to wait 
till runtime to figure out what we broke!

The following also introduces the concept of inversion of control when
registering a facade. Instead of the all facades file knowing about each
and every version of the facade, we invert that. Now we ask the package
what it would like to register and it is up to that package what it's
registering. We no longer require exported functions to be randomly
exported. It is more considered. This reduces the footprint of facades
public API. The downside for now, that will need to be improved is the
usage of export_test.go files. A lot of the tests are functional/behavioural 
tests, when they should be unit tests (both have their place).

The advantage to doing all this work, is that shimming a API facade layer
in front will likely be a lot simpler as reflection isn't so extensively
used.


## QA steps

Ensure when rebuilding the schema we didn't break anything. That
outputs the latest facades in json format and we can diff that output
against the tip of 2.9 to see that indeed nothing changed.

```sh
$ make rebuild-schema
$ diff <2.9 schema.json> <new schema.json>
```

Run some tests to ensure we didn't break anything.

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu
$ juju status
```
